File: C++/Class_Work/week4/assignment4a/assignment4a.cpp
//Priceton Braswell
//Assignment 4A 



#include <cmath>
#include <string>
#include <unordered_map>
#include <vector>
#include <iostream>



//---------------------------------------------------------------------------------------------------------
//---------------------------------------------------------------------------------------------------------


class Person {
private:
  std::string m_sex;
  std::string m_first_name;
  int m_weight_class_id;   // overweight = 0, probably not overweight = 1, normal weight = 2, under weight = 3
  int m_height_feet;
  int m_height_inches;
  int m_body_fat_percent;
  int m_id;
  float m_weight_pounds;
  float m_bmi;


  void m_calculate_bmi(int feet, int inches, float pounds) {
    m_bmi = pounds / ((feet * 12 + inches) * (feet * 12 + inches)) * 703;
  }

   void m_calculate_the_fatness(float bmi, int body_fat_percent, std::string sex) {
     if (bmi > 25) {
       if (sex == "male" && body_fat_percent > 24) {
         m_weight_class_id = 0;
         return;
       }
       if (sex == "female" && body_fat_percent > 31) {
        m_weight_class_id = 0;
         return;
       }
       m_weight_class_id = 1;
       return;
     }
     if (bmi > 18.5) {
       m_weight_class_id = 2;
       return;
     }
     m_weight_class_id = 3;
     return;
   }






public:
  Person(std::string name, int height_inches, int height_feet, std::string sex, float pounds, int fat_percent, int id)
    : m_first_name(name), 
      m_height_inches(height_inches), 
      m_height_feet(height_feet),
      m_sex(sex), 
      m_weight_pounds(pounds),
      m_body_fat_percent(fat_percent),
      m_id(id){
    m_calculate_bmi(m_height_feet, m_height_inches, m_weight_pounds);
    m_calculate_the_fatness(m_bmi, m_body_fat_percent, m_sex);
  }

  std::string get_sex() {
    return m_sex;
  }

  std::string get_first_name() {
    return m_first_name;
  }

  int get_height_feet() {
    return m_height_feet;
  }

  int get_height_inches() {
    return m_height_inches;
  }

  int get_body_fat_percent() {
    return m_body_fat_percent;
  }

  int get_person_id() {
    return m_id;
  }

  float get_weight() {
    return m_weight_pounds;
  }

  float get_bmi() {
    return m_bmi;
  }

  int get_weight_class_id() {
    return m_weight_class_id;
  }


};

//---------------------------------------------------------------------------------------------------------
//---------------------------------------------------------------------------------------------------------







class Group {
private:


  int m_group_count;
  int m_male_count;
  int m_female_count;
  float m_group_bmi_avg;
  float m_male_bmi_avg;
  float m_female_bmi_avg;
  std::vector<Person> m_people;
  std::unordered_map<int, int> m_weight_class_counts = {{0,0}, {1,0}, {2,0}, {3,0}};   // weight_class_id : number of people in that weight class 
  std::string m_weight_class_names[4] = {"overweight", "probably not overweight", "normal", "underweight" };  

  

//I HAD CHATGPT ADD IF STATEMENTS IN FUNCTION BELOW AND STATIC CAST TO HANDLE FLOAT EXCEPTION ERROR FOR DIVIDING BY ZERO. 
void m_print_weight_class_percentages() {
    if (m_group_count != 0) {
        if (m_weight_class_counts[0] != 0) {
            std::cout << "overweight members: " 
                      << (static_cast<float>(m_weight_class_counts[0]) / m_group_count) * 100 
                      << "% \n";
        } else {
            std::cout << "overweight members: 0% \n";
        }

        if (m_weight_class_counts[1] != 0) {
            std::cout << "probably not overweight members: " 
                      << (static_cast<float>(m_weight_class_counts[1]) / m_group_count) * 100 
                      << "% \n";
        } else {
            std::cout << "probably not overweight members: 0% \n";
        }

        if (m_weight_class_counts[2] != 0) {
            std::cout << "normal members: " 
                      << (static_cast<float>(m_weight_class_counts[2]) / m_group_count) * 100 
                      << "% \n";
        } else {
            std::cout << "normal members: 0% \n";
        }

        if (m_weight_class_counts[3] != 0) {
            std::cout << "underweight members: " 
                      << (static_cast<float>(m_weight_class_counts[3]) / m_group_count) * 100 
                      << "% \n";
        } else {
            std::cout << "underweight members: 0% \n";
        }
    } else {
        std::cout << "Group count is zero. Cannot calculate percentages.\n";
    }

    return;
}


  void m_print_person_info(Person person) {
    
    std::cout << "Name: " << person.get_first_name() << "\n";
    std::cout << "Height: " << person.get_height_feet() << "," << person.get_height_inches() << "\n";
    std::cout << "Weight: " << person.get_weight() << "\n";
    std::cout << "Body Fat Percentage: " << person.get_body_fat_percent() << "\n";
    std::cout << "Sex: " << person.get_sex() << "\n";
    std::cout << "BMI: " << person.get_bmi() << "\n";
    std::cout << "Weight Class: " << m_weight_class_names[person.get_weight_class_id()] << "\n";
  }

//-------------------------------------------------------------------------------------
  // I HAD CHAT GPT DO THE THREE CALCULATE FUNCTIONS BELOW. 
  // NOT BECAUSE I COULDNT DO IT. BUT BECAUSE ITS 2AM AND IM TIRED.
  // I REMEMBER IN THE FIRST VIDEO YOU SAID TO LET YOU KNOW WHEN WE DO THAT. 

  void m_calculate_group_bmi_avg() {
    if (m_group_count == 0) {
      m_group_bmi_avg = 0;  // No people, so average BMI is 0
      return;
    }

    float total_bmi = 0;
    for (Person person : m_people) {
      total_bmi += person.get_bmi();
    }

    m_group_bmi_avg = total_bmi / m_group_count;
  }

  // Function to calculate the average BMI for male members of the group
  void m_calculate_male_bmi_avg() {
    if (m_male_count == 0) {
      m_male_bmi_avg = 0;  // No males in the group
      return;
    }

    float total_bmi_males = 0;
    for (Person person : m_people) {
      if (person.get_sex() == "male") {
        total_bmi_males += person.get_bmi();
      }
    }

    m_male_bmi_avg = total_bmi_males / m_male_count;
  }

  // Function to calculate the average BMI for female members of the group
  void m_calculate_female_bmi_avg() {
    if (m_female_count == 0) {
      m_female_bmi_avg = 0;  // No females in the group
      return;
    }

    float total_bmi_females = 0;
    for (Person person : m_people) {
      if (person.get_sex() == "female") {
        total_bmi_females += person.get_bmi();
      }
    }

    m_female_bmi_avg = total_bmi_females / m_female_count;
  }


public:

  Group()
   :m_group_count(0),
    m_male_count(0),
    m_female_count(0),
    m_male_bmi_avg(0),
    m_female_bmi_avg(0),
    m_group_bmi_avg(0) {}


  void print_group_bmi_for_4a() {
    std::cout << "Average BMI for group: " << m_group_bmi_avg << "\n";
  }


  void print_weight_class_percentages_for_4a() {
    m_print_weight_class_percentages();
  }

  void print_group_info() {
    if (m_group_count == 0) {
      std::cout << "there are no people in the group \n";
      return;
    }

    std::cout << "Number of people in group: " << m_group_count << "\n";
    std::cout << "Group BMI average: " << m_group_bmi_avg << "\n";
    std::cout << "Male BMI average: " << m_male_bmi_avg << "\n" ;
    std::cout << "Female BMI average: " << m_female_bmi_avg << "\n";
    m_print_weight_class_percentages();
  }

  void print_person_info(int person_id) {
    if (person_id > m_group_count - 1) {
      std::cout << "UNABLE TO FIND PERSON ID " << person_id;
    }
    m_print_person_info(m_people[person_id]);
  }



  void add_person() {
    std::string name;
    int feet;
    int inches;
    float pounds;
    int body_fat_percent;
    std::string sex;
    int person_id;

    
    std::cout << "please enter your first name "  << "\n";
    std::cin >> name; 

    std::cout << "please enter your height in feet "  << "\n";
    std::cin >> feet ; 

    std::cout << "how many inches on top of those feet? "  << "\n";
    std::cin >> inches; 

    std::cout << "please enter how many pounds you are "  << "\n";
    std::cin >> pounds; 

    std::cout << "please enter your body fat percent "  << "\n";
    std::cin >> body_fat_percent; 

    std::cout << "male or female? "  << "\n";
    std::cin >> sex; 

    while (sex != "male" && sex != "female") {
      std::cout << "Try again. Type male or female all lower case and press enter. \n";
      std::cin >> sex;

    }

    person_id = m_group_count;
 
    Person person_to_add = Person(name, inches, feet, sex, pounds, body_fat_percent, person_id);
    m_people.push_back(person_to_add);
    m_weight_class_counts[person_to_add.get_weight_class_id()]++;



    m_group_count += 1;
    if (sex == "male") {
      m_male_count += 1;
      m_calculate_male_bmi_avg();
      
      
    }
    if (sex == "female") {
      m_female_count +=1;
      m_calculate_female_bmi_avg();
    }
    m_calculate_group_bmi_avg();


    std::cout << "PERSON CREATED! PERSON ID IS " << person_id << "\n";
  }

};




int main (int argc, char *argv[]) {
  Group group = Group();

  for (int i=0; i < 5; i++ ) {
    group.add_person();
  }
  group.print_group_bmi_for_4a();
  group.print_weight_class_percentages_for_4a();

  return 0;
}

File: Java_Class/week4/assignment4a/Assignment4a.java
import java.util.*;

class Person {
    private String sex;
    private String firstName;
    private int weightClassId; // overweight = 0, probably not overweight = 1, normal weight = 2, underweight = 3
    private int heightFeet;
    private int heightInches;
    private int bodyFatPercent;
    private int id;
    private float weightPounds;
    private float bmi;

    private void calculateBmi(int feet, int inches, float pounds) {
        this.bmi = pounds / ((feet * 12 + inches) * (feet * 12 + inches)) * 703;
    }

    private void calculateFatness(float bmi, int bodyFatPercent, String sex) {
        if (bmi > 25) {
            if (sex.equals("male") && bodyFatPercent > 24) {
                this.weightClassId = 0;
                return;
            }
            if (sex.equals("female") && bodyFatPercent > 31) {
                this.weightClassId = 0;
                return;
            }
            this.weightClassId = 1;
            return;
        }
        if (bmi > 18.5) {
            this.weightClassId = 2;
            return;
        }
        this.weightClassId = 3;
    }

    public Person(String name, int heightInches, int heightFeet, String sex, float pounds, int fatPercent, int id) {
        this.firstName = name;
        this.heightInches = heightInches;
        this.heightFeet = heightFeet;
        this.sex = sex;
        this.weightPounds = pounds;
        this.bodyFatPercent = fatPercent;
        this.id = id;
        calculateBmi(this.heightFeet, this.heightInches, this.weightPounds);
        calculateFatness(this.bmi, this.bodyFatPercent, this.sex);
    }

    public String getSex() {
        return sex;
    }

    public String getFirstName() {
        return firstName;
    }

    public int getHeightFeet() {
        return heightFeet;
    }

    public int getHeightInches() {
        return heightInches;
    }

    public int getBodyFatPercent() {
        return bodyFatPercent;
    }

    public int getPersonId() {
        return id;
    }

    public float getWeight() {
        return weightPounds;
    }

    public float getBmi() {
        return bmi;
    }

    public int getWeightClassId() {
        return weightClassId;
    }
}

class Group {
    private int groupCount;
    private int maleCount;
    private int femaleCount;
    private float groupBmiAvg;
    private float maleBmiAvg;
    private float femaleBmiAvg;
    private List<Person> people = new ArrayList<>();
    private Map<Integer, Integer> weightClassCounts = new HashMap<>();
    private String[] weightClassNames = {"overweight", "probably not overweight", "normal", "underweight"};

    public Group() {
        groupCount = 0;
        maleCount = 0;
        femaleCount = 0;
        maleBmiAvg = 0;
        femaleBmiAvg = 0;
        groupBmiAvg = 0;
        weightClassCounts.put(0, 0);
        weightClassCounts.put(1, 0);
        weightClassCounts.put(2, 0);
        weightClassCounts.put(3, 0);
    }

    private void printWeightClassPercentages() {
        if (groupCount != 0) {
            for (int i = 0; i < 4; i++) {
                int count = weightClassCounts.get(i);
                if (count != 0) {
                    System.out.println(weightClassNames[i] + " members: " + ((float) count / groupCount) * 100 + "%");
                } else {
                    System.out.println(weightClassNames[i] + " members: 0%");
                }
            }
        } else {
            System.out.println("Group count is zero. Cannot calculate percentages.");
        }
    }

    private void printPersonInfo(Person person) {
        System.out.println("Name: " + person.getFirstName());
        System.out.println("Height: " + person.getHeightFeet() + "," + person.getHeightInches());
        System.out.println("Weight: " + person.getWeight());
        System.out.println("Body Fat Percentage: " + person.getBodyFatPercent());
        System.out.println("Sex: " + person.getSex());
        System.out.println("BMI: " + person.getBmi());
        System.out.println("Weight Class: " + weightClassNames[person.getWeightClassId()]);
    }

    private void calculateGroupBmiAvg() {
        if (groupCount == 0) {
            groupBmiAvg = 0;
            return;
        }

        float totalBmi = 0;
        for (Person person : people) {
            totalBmi += person.getBmi();
        }

        groupBmiAvg = totalBmi / groupCount;
    }

    private void calculateMaleBmiAvg() {
        if (maleCount == 0) {
            maleBmiAvg = 0;
            return;
        }

        float totalBmiMales = 0;
        for (Person person : people) {
            if (person.getSex().equals("male")) {
                totalBmiMales += person.getBmi();
            }
        }

        maleBmiAvg = totalBmiMales / maleCount;
    }

    private void calculateFemaleBmiAvg() {
        if (femaleCount == 0) {
            femaleBmiAvg = 0;
            return;
        }

        float totalBmiFemales = 0;
        for (Person person : people) {
            if (person.getSex().equals("female")) {
                totalBmiFemales += person.getBmi();
            }
        }

        femaleBmiAvg = totalBmiFemales / femaleCount;
    }
  
    public void printGroupBMIfor4a() {
      System.out.println("Average BMI for group: " + groupBmiAvg);
    }
  
  
    public void printWeightClassPercentageFor4a() {
      printWeightClassPercentages();
    }


    public void printGroupInfo() {
        if (groupCount == 0) {
            System.out.println("There are no people in the group.");
            return;
        }

        System.out.println("Number of people in group: " + groupCount);
        System.out.println("Group BMI average: " + groupBmiAvg);
        System.out.println("Male BMI average: " + maleBmiAvg);
        System.out.println("Female BMI average: " + femaleBmiAvg);
        printWeightClassPercentages();
    }

    public void printPersonInfo(int personId) {
        if (personId > groupCount - 1) {
            System.out.println("UNABLE TO FIND PERSON ID " + personId);
            return;
        }
        printPersonInfo(people.get(personId));
    }

    public void addPerson() {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Please enter your first name:");
        String name = scanner.nextLine();

        System.out.println("Please enter your height in feet:");
        int feet = scanner.nextInt();

        System.out.println("How many inches on top of those feet?");
        int inches = scanner.nextInt();

        System.out.println("Please enter how many pounds you are:");
        float pounds = scanner.nextFloat();

        System.out.println("Please enter your body fat percent:");
        int bodyFatPercent = scanner.nextInt();

        System.out.println("Male or female?");
        String sex = scanner.next();

        while (!sex.equals("male") && !sex.equals("female")) {
            System.out.println("Try again. Type male or female all lowercase and press enter.");
            sex = scanner.next();
        }

        int personId = groupCount;

        Person personToAdd = new Person(name, inches, feet, sex, pounds, bodyFatPercent, personId);
        people.add(personToAdd);
        weightClassCounts.put(personToAdd.getWeightClassId(), weightClassCounts.get(personToAdd.getWeightClassId()) + 1);

        groupCount += 1;
        if (sex.equals("male")) {
            maleCount += 1;
            calculateMaleBmiAvg();
        }
        if (sex.equals("female")) {
            femaleCount += 1;
            calculateFemaleBmiAvg();
        }
        calculateGroupBmiAvg();

        System.out.println("PERSON CREATED! PERSON ID IS " + personId);
    }
}

public class Assignment4a {
    public static void main(String[] args) {
        Group group = new Group();
        Scanner scanner = new Scanner(System.in);
        boolean running = true;
        for (int i=0; i < 5; i++ ) {
          group.addPerson();
        }
        group.printGroupBMIfor4a();
        group.printWeightClassPercentageFor4a();
      
    }
}
